src/main/java/htsjdk/samtools/cram/structure/CRAMRecord.java

```diff
@@ -0,0 +1,927 @@
+        ValidationUtils.validateArg(sequentialIndex >= 0, "sequential index must be >= 0");
+        ValidationUtils.nonNull(readGroupMap);
+
+        this.sequentialIndex = sequentialIndex;
+
+        // default to detached state until the actual mate information state is resolved during mate normalization
+        setToDetachedState();
+
+        // flags
+        setMultiFragment(samRecord.getReadPairedFlag());
+        setProperPair(samRecord.getReadPairedFlag() && samRecord.getProperPairFlag());
+        setSegmentUnmapped(samRecord.getReadUnmappedFlag());
+        setFirstSegment(samRecord.getReadPairedFlag() && samRecord.getFirstOfPairFlag());
+        setLastSegment(samRecord.getReadPairedFlag() && samRecord.getSecondOfPairFlag());
+        setNegativeStrand(samRecord.getReadNegativeStrandFlag());
+        setSecondaryAlignment(samRecord.isSecondaryAlignment());
+        setSupplementary(samRecord.getSupplementaryAlignmentFlag());
+        setVendorFiltered(samRecord.getReadFailsVendorQualityCheckFlag());
+        setDuplicate(samRecord.getDuplicateReadFlag());
+
+        readName = samRecord.getReadName();
+        referenceIndex = samRecord.getReferenceIndex();
+
+        readLength = samRecord.getReadLength();
+        alignmentStart = samRecord.getAlignmentStart();
+        if (samRecord.getReadUnmappedFlag()) {
+            readFeatures = new CRAMRecordReadFeatures();
+            alignmentEnd = AlignmentContext.NO_ALIGNMENT_END;
+        } else {
+            readFeatures = new CRAMRecordReadFeatures(samRecord, referenceBases);
+            alignmentEnd = readFeatures.getAlignmentEnd(alignmentStart, readLength);
+        }
+
+        templateSize = samRecord.getInferredInsertSize();
+        mappingQuality = samRecord.getMappingQuality();
+
+        if (samRecord.getReadPairedFlag()) {
+            mateAlignmentStart = samRecord.getMateAlignmentStart();
+            setMateUnmapped(samRecord.getMateUnmappedFlag());
+            setMateNegativeStrand(samRecord.getMateNegativeStrandFlag());
+            mateReferenceIndex = samRecord.getMateReferenceIndex();
+        } else {
+            mateAlignmentStart = 0;
+            mateReferenceIndex = SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
+        }
+
+        // CRAM read bases are limited to ACGTN, see https://github.com/samtools/hts-specs/blob/master/CRAMv3.pdf
+        // passage 10.2 on read bases. However, BAM format allows upper case IUPAC codes without a dot, so we
+        // follow the same approach to reproduce the behaviour of samtools.
+        // copy read bases to avoid changing the original record:
+        final byte[] bases = samRecord.getReadBases();
+        readBases = bases == null || bases.equals(SAMRecord.NULL_SEQUENCE) ?
+                    SAMRecord.NULL_SEQUENCE :
+                    SequenceUtil.toBamReadBasesInPlace(Arrays.copyOf(bases, samRecord.getReadLength()));
```

Comment:
Not sure which "change" you're referring as far as the spec; it could probably be more explicit, but I think this is the same transformation done for BAM, with which CRAM claims compatibility.

Reply:
There are IUPAC round trip test in `CRAMComplianceTest.ambiguityCodeVerificationTest` that IIRC tests all ambiguity codes.